src/main/java/de/tum/in/www1/artemis/domain/Result.java

```diff
@@ -485,27 +485,59 @@ public void filterSensitiveInformation() {
      * @param removeHiddenFeedback if feedbacks marked with visibility 'after due date' should also be removed.
      */
     public void filterSensitiveFeedbacks(boolean removeHiddenFeedback) {
-        var filteredFeedback = createFilteredFeedbacks(removeHiddenFeedback);
+        filterSensitiveFeedbacks(removeHiddenFeedback, participation.getExercise());
+    }
+
+    /**
+     * Removes all feedback details that should not be passed to the student.
+     *
+     * @param removeHiddenFeedback if feedbacks marked with visibility 'after due date' should also be removed.
+     * @param exercise             the exercise related to this result. Used to determine if test case names should be removed.
+     */
+    public void filterSensitiveFeedbacks(boolean removeHiddenFeedback, Exercise exercise) {
+        var filteredFeedback = createFilteredFeedbacks(removeHiddenFeedback, exercise);
         setFeedbacks(filteredFeedback);
 
-        // TODO: this is not good code!
+        if (exercise instanceof ProgrammingExercise) {
+            updateTestCaseCount();
+        }
+    }
+
+    /**
+     * Updates the testCaseCount and passedTestCaseCount attributes after filtering the feedback.
+     */
+    private void updateTestCaseCount() {
         var testCaseFeedback = feedbacks.stream().filter(Feedback::isTestFeedback).toList();
+
+        // TODO: this is not good code!
         setTestCaseCount(testCaseFeedback.size());
         setPassedTestCaseCount((int) testCaseFeedback.stream().filter(feedback -> Boolean.TRUE.equals(feedback.isPositive())).count());
     }
 
     /**
      * Returns a new list that only contains feedback that should be passed to the student.
      * Does not change the feedbacks attribute of this entity.
+     * Also removes the test names from all feedback if it should not be shown to the student.
      *
      * @see ResultDTO
      *
      * @param removeHiddenFeedback if feedbacks marked with visibility 'after due date' should also be removed.
+     * @param exercise             used to check if students can see the test case names
      * @return the new filtered list
      */
-    public List<Feedback> createFilteredFeedbacks(boolean removeHiddenFeedback) {
-        return feedbacks.stream().filter(feedback -> !feedback.isInvisible()).filter(feedback -> !removeHiddenFeedback || !feedback.isAfterDueDate())
+    public List<Feedback> createFilteredFeedbacks(boolean removeHiddenFeedback, Exercise exercise) {
+        var filteredFeedback = feedbacks.stream().filter(feedback -> !feedback.isInvisible()).filter(feedback -> !removeHiddenFeedback || !feedback.isAfterDueDate())
                 .collect(Collectors.toCollection(ArrayList::new));
+
+        if (exercise instanceof ProgrammingExercise programmingExercise && !Boolean.TRUE.equals(programmingExercise.getShowTestNamesToStudents())) {
```

Comment:
this getter returns a Boolean but the underlying variable is a primitive type, can we change the getter and remove the Boolean.TRUE.equals

Reply:
but the table column is nullable... I'll just keep that for now, maybe we can migrate a for of these Boolean-Attributes in a followup.